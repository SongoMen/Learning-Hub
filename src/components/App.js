import React, { Component } from "react";
import {
  Route,
  BrowserRouter as Router,
  Redirect,
  Switch
} from "react-router-dom";
import { firebaseAuth } from "./auth";

import Loader from "./elements/Loader";
import LandingPage from "./LandingPage/LandingPage";
import Login from "./Login/Login";
import Register from "./Register/Register";
import Dashboard from "./Dashboard/Dashboard";
import CourseContainer from "./CoursePage/CourseContainer";

function PrivateRoute({ component: Component, authed, ...rest }) {
  return (
    <Route
      {...rest}
      render={props =>
        authed === true ? (
          <Component {...props} {...rest} />
        ) : (
          <Redirect
            to={{ pathname: "/login", state: { from: props.location } }}
          />
        )
      }
    />
  );
}

function PublicRoute({ component: Component, authed, ...rest }) {
  return (
    <Route
      {...rest}
      render={props =>
        authed === false ? (
          <Component {...props} />
        ) : (
          <Redirect to="/dashboard" />
        )
      }
    />
  );
}

class App extends Component {
  _isMounted = false;
  state = {
    authed: false,
    loading: true,
    theme: ""
  };
  componentDidMount() {
    this._isMounted = true;
    this.removeListener = firebaseAuth().onAuthStateChanged(user => {
      if (this._isMounted) {
        if (user) {
          this.setState({
            authed: true,
            loading: false
          });
        } else {
          this.setState({
            authed: false,
            loading: false
          });
        }
      }
    });
  }
  componentWillUnmount() {
    this._isMounted = false;
    this.removeListener();
  }

  render() {
    return this.state.loading ? (
      <Loader />
    ) : (
      <Router>
        <div className="container">
          <Switch>
            <Route exact path="/" component={LandingPage} />
            <PublicRoute
              authed={this.state.authed}
              path="/login"
              component={Login}
            />
            <PublicRoute
              authed={this.state.authed}
              path="/register"
              component={Register}
            />
            <PrivateRoute
              authed={this.state.authed}
              path="/dashboard"
              component={Dashboard}
            />
            <PrivateRoute
              authed={this.state.authed}
              path="/courses"
              component={Dashboard}
            />
            <PrivateRoute
              authed={this.state.authed}
              path="/devpanel"
              component={Dashboard}
            />
            <PrivateRoute
              authed={this.state.authed}
              path="/course/:name"
              component={CourseContainer}
              exact
              name={window.location.pathname.split("/")[2]}
            />
            <PrivateRoute
              authed={this.state.authed}
              path="/course/:name/:id"
              component={CourseContainer}
              id={window.location.pathname.split("/")[3]}
            />
          </Switch>
        </div>
      </Router>
    );
  }
}

export default App;
