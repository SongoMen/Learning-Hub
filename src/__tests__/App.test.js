import React from "react";
import App from "../components/App";
import renderer from "react-test-renderer";

describe("App", () => {
  it("renders", () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
