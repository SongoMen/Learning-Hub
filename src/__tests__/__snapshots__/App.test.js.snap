// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<div
  className="loader"
>
  <svg>
    <defs>
      <filter
        id="goo"
      >
        <feGaussianBlur
          in="SourceGraphic"
          result="blur"
          stdDeviation="2"
        />
        <feColorMatrix
          in="blur"
          mode="matrix"
          result="gooey"
          values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 5 -2"
        />
        <feComposite
          in="SourceGraphic"
          in2="gooey"
          operator="atop"
        />
      </filter>
    </defs>
  </svg>
</div>
`;

exports[`Button component Matches the snapshot 1`] = `
<div
  className="loader"
>
  <svg>
    <defs>
      <filter
        id="goo"
      >
        <feGaussianBlur
          in="SourceGraphic"
          result="blur"
          stdDeviation="2"
        />
        <feColorMatrix
          in="blur"
          mode="matrix"
          result="gooey"
          values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 5 -2"
        />
        <feComposite
          in="SourceGraphic"
          in2="gooey"
          operator="atop"
        />
      </filter>
    </defs>
  </svg>
</div>
`;
